iothub/device/src/InternalClient.cs

```diff
@@ -1071,37 +1072,39 @@ public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCal
 
         /// <summary>
         /// Set a callback that will be called whenever the client receives a state update
-        /// (desired or reported) from the service.  This has the side-effect of subscribing
-        /// to the PATCH topic on the service.
+        /// (desired or reported) from the service.
+        /// Set callback value to null to clear.
         /// </summary>
+        /// <remarks>
+        /// This has the side-effect of subscribing to the PATCH topic on the service.
+        /// </remarks>
         /// <param name="callback">Callback to call after the state update has been received and applied</param>
         /// <param name="userContext">Context object that will be passed into callback</param>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
         public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCallback callback, object userContext, CancellationToken cancellationToken)
         {
-            if (callback == null)
-            {
-                throw new ArgumentNullException(nameof(callback));
-            }
-
             // Codes_SRS_DEVICECLIENT_18_003: `SetDesiredPropertyUpdateCallbackAsync` shall call the transport to register for PATCHes on it's first call.
             // Codes_SRS_DEVICECLIENT_18_004: `SetDesiredPropertyUpdateCallbackAsync` shall not call the transport to register for PATCHes on subsequent calls
-            if (!_twinPatchSubscribedWithService)
+            try
             {
-                try
+                if (callback != null && !_twinPatchSubscribedWithService)
                 {
                     await InnerHandler.EnableTwinPatchAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
shouldn't you grab a lock before enabling the twin patches/ disabling them? If you have parallel calls that both set and unset the callback, what is the internal state going to be then?

Reply:
Agree. It is something I have thought about, not addressing in this PR though. Let's discuss this offline.

Reply:
Sure, we can discuss this offline.

Reply:
Addressed this.